syntax = "proto3";
package hivekernel.core;

option go_package = "github.com/selemilka/hivekernel/api/proto/hivepb";

import "agent.proto";

// ─── CoreService — provided by core, called by agents ───

service CoreService {
  // Process management
  rpc SpawnChild(hivekernel.agent.SpawnRequest) returns (hivekernel.agent.SpawnResponse);
  rpc KillChild(hivekernel.agent.KillRequest) returns (hivekernel.agent.KillResponse);
  rpc GetProcessInfo(ProcessInfoRequest) returns (ProcessInfo);
  rpc ListChildren(ListChildrenRequest) returns (ListChildrenResponse);

  // IPC
  rpc SendMessage(hivekernel.agent.SendMessageRequest) returns (hivekernel.agent.SendMessageResponse);
  rpc Subscribe(SubscribeRequest) returns (stream hivekernel.agent.AgentMessage);

  // Shared memory
  rpc StoreArtifact(hivekernel.agent.StoreArtifactRequest) returns (hivekernel.agent.StoreArtifactResponse);
  rpc GetArtifact(hivekernel.agent.GetArtifactRequest) returns (hivekernel.agent.GetArtifactResponse);
  rpc ListArtifacts(ListArtifactsRequest) returns (ListArtifactsResponse);

  // Resources
  rpc GetResourceUsage(ResourceUsageRequest) returns (ResourceUsage);
  rpc RequestResources(ResourceRequest) returns (ResourceResponse);

  // Escalation
  rpc Escalate(hivekernel.agent.EscalateRequest) returns (hivekernel.agent.EscalateResponse);

  // Logging
  rpc Log(hivekernel.agent.LogRequest) returns (hivekernel.agent.LogResponse);
  rpc ReportMetric(MetricRequest) returns (MetricResponse);

  // Task execution
  rpc ExecuteTask(ExecuteTaskRequest) returns (ExecuteTaskResponse);

  // Event sourcing
  rpc SubscribeEvents(SubscribeEventsRequest) returns (stream ProcessEvent);

  // Cron management
  rpc AddCron(AddCronRequest) returns (AddCronResponse);
  rpc RemoveCron(RemoveCronRequest) returns (RemoveCronResponse);
  rpc ListCron(ListCronRequest) returns (ListCronResponse);

  // Inbox inspection
  rpc ListInbox(ListInboxRequest) returns (ListInboxResponse);

  // Sibling / wait support (mirror of syscall equivalents)
  rpc ListSiblings(hivekernel.agent.ListSiblingsRequest) returns (hivekernel.agent.ListSiblingsResponse);
  rpc WaitChild(hivekernel.agent.WaitChildRequest) returns (hivekernel.agent.WaitChildResponse);
}

// ─── Process management ───

message ProcessInfoRequest {
  uint64 pid = 1;
}

message ProcessInfo {
  uint64 pid = 1;
  uint64 ppid = 2;
  string user = 3;
  string name = 4;
  hivekernel.agent.AgentRole role = 5;
  hivekernel.agent.CognitiveTier cognitive_tier = 6;
  string model = 7;
  hivekernel.agent.AgentState state = 8;
  string vps = 9;
  uint64 tokens_consumed = 10;
  float context_usage_percent = 11;
  uint32 child_count = 12;
  uint64 started_at = 13;
  string current_task_id = 14;
  string runtime_addr = 15;
}

message ListChildrenRequest {
  bool recursive = 1;
  hivekernel.agent.AgentState filter_state = 2;
}

message ListChildrenResponse {
  repeated ProcessInfo children = 1;
}

// ─── IPC ───

message SubscribeRequest {
  string queue = 1;
  string type_filter = 2;
  hivekernel.agent.Priority min_priority = 3;
}

// ─── Shared memory ───

message ListArtifactsRequest {
  string prefix = 1;
  uint64 stored_by_pid = 2;
}

message ListArtifactsResponse {
  repeated ArtifactMeta artifacts = 1;
}

message ArtifactMeta {
  string artifact_id = 1;
  string key = 2;
  string content_type = 3;
  uint64 size_bytes = 4;
  uint64 stored_by_pid = 5;
  uint64 stored_at = 6;
}

// ─── Resources ───

message ResourceUsageRequest {}

message ResourceUsage {
  uint64 tokens_consumed = 1;
  uint64 tokens_remaining = 2;
  uint32 children_active = 3;
  uint32 children_max = 4;
  float context_usage_percent = 5;
  uint64 uptime_seconds = 6;
}

message ResourceRequest {
  string resource_type = 1;
  uint64 amount = 2;
  string justification = 3;
}

message ResourceResponse {
  bool granted = 1;
  uint64 granted_amount = 2;
  string reason = 3;
}

// ─── Logging ───

message MetricRequest {
  string name = 1;
  double value = 2;
  map<string, string> labels = 3;
}

message MetricResponse {}

// ─── Task execution ───

message ExecuteTaskRequest {
  uint64 target_pid = 1;
  string description = 2;
  map<string, string> params = 3;
  int32 timeout_seconds = 4;
}

message ExecuteTaskResponse {
  bool success = 1;
  hivekernel.agent.TaskResult result = 2;
  string error = 3;
}

// ─── Event sourcing ───

message ProcessEvent {
  uint64 seq = 1;
  int64  timestamp_ms = 2;
  string type = 3;       // "spawned", "state_changed", "removed", "log", "llm_call", "tool_call", "cron_executed"
  uint64 pid = 4;
  uint64 ppid = 5;
  string name = 6;
  string role = 7;
  string tier = 8;
  string model = 9;
  string state = 10;
  string old_state = 11;
  string new_state = 12;
  string level = 13;
  string message = 14;
  string reply_to = 15;
  string payload_preview = 16;
  string trace_id = 17;
  string trace_span = 18;
  string message_id = 19;
  map<string, string> fields = 20;  // structured metadata for typed events
}

message SubscribeEventsRequest {
  uint64 since_seq = 1;  // 0 = from now, >0 = replay from seq
}

// ─── Cron management ───

message CronEntryProto {
  string id = 1;
  string name = 2;
  string cron_expression = 3;
  string action = 4;              // "spawn", "wake", "execute"
  uint64 target_pid = 5;
  string execute_description = 6;
  map<string, string> execute_params = 7;
  bool enabled = 8;
  int64 last_run_ms = 9;
  int64 next_run_ms = 10;
  int32 last_exit_code = 11;
  string last_output = 12;
  int64 last_duration_ms = 13;
}

message AddCronRequest {
  string name = 1;
  string cron_expression = 2;
  string action = 3;              // "execute" (default), "spawn", "wake"
  uint64 target_pid = 4;
  string execute_description = 5;
  map<string, string> execute_params = 6;
}

message AddCronResponse {
  string cron_id = 1;
  string error = 2;
}

message RemoveCronRequest {
  string cron_id = 1;
}

message RemoveCronResponse {
  bool ok = 1;
  string error = 2;
}

message ListCronRequest {}

message ListCronResponse {
  repeated CronEntryProto entries = 1;
}

// ─── Inbox inspection ───

message ListInboxRequest {
  uint64 pid = 1;
}

message ListInboxResponse {
  repeated hivekernel.agent.AgentMessage messages = 1;
  int32 total = 2;
}
